libsolidity/ast/Types.h

```diff
@@ -1598,6 +1598,7 @@ class TypeType: public Type
 	bool hasSimpleZeroValueInMemory() const override { solAssert(false, ""); }
 	std::string toString(bool _withoutDataLocation) const override { return "type(" + m_actualType->toString(_withoutDataLocation) + ")"; }
 	MemberList::MemberMap nativeMembers(ASTNode const* _currentScope) const override;
+	Type const* mobileType() const override { return nullptr; }
```

Comment:
I'm still confused whether this PR is meant to fix only `abi.decode()` or more :) In https://github.com/ethereum/solidity/pull/14347#issuecomment-1609817267 you confirmed it's only `abi.decide()` and thats the only thing you have tests for, but I think this change is more general and covers more cases.

Please confirm is any of these is now forbidden (I mean in general, not just in `abi.decode()`), and if so, make sure to verify it with a test case:
- Explicit type conversions (`c ? uint : uint)(0)`).
- `type` (`type(true ? uint : uint).max`).
- Statements that consist of type only (`c ? uint : uint;`)
    - Base type of an array in such a statement (`(c ? uint[2] : uint[2])[3];`).
- Contracts (`(c ? C : C).f()`).
    - Including libraries (`(c ? L : L).f()`).
- Struct constructors (`(c ? S : S)(x, y, z)`).

Also, the changelog entry should be updated if you're fixing any of them here.

Reply:
Yeah, this got a little bit confusing because originally the changes made were targeting `abi.decode` specifically. Later when we changed to the `mobileType` override, it became more general, but in my mind I still considered that this was only about `abi.decode` and we would push other tests in another PR.
I will go over all the general cases and add them.